LM-3: file export leave list

rgillen-uno opened the pull request:
Jira issue: #LM-3
Implementation of the CSV data export option.
Main changes include:
- DataETL project
- WebApi mappings
- Event Controller
- CsvService
- Dependency Injection bindings

Currently only implemented on one endpoint.
Example Usage:
findByEventStatus/{eventStatusId}/{eventTypeId}/csv
http://localhost:8081/event/findByEventStatus/1/1/csv

rgillen-uno committed:
Add project and structure.
Integrate into event service and event controller.
Define new mappings.

rgillen-uno committed:
Add new common attributes and ETL (Extract, Transform, Load) project for converting data to CSV format.
Add several converters to help with conversion of data to string type.
Add new mappings for EventDto to new EventDataTransformModel.
Remove project wide unused using statements.

rgillen-uno committed:
Add comments to ICsvService.

kkuczkowski-unosq commented on AdminCore.DataETL/Attributes/TypeConverterAttribute.cs:
```diff
@@ -0,0 +1,16 @@
+    [AttributeUsage(AttributeTargets.Property)]
+    public class TypeConverterAttribute : Attribute
+    {
+//        public ITypeConverter Converter { get; set; }
```
clean this up?

BrownMark replied:
Just a reminder to remove the commented code.

kkuczkowski-unosq commented on AdminCore.DataETL/CsvChoEtlAdapter.cs:
```diff
@@ -0,0 +1,117 @@
+        
+        /// <summary>
+        /// Creates an ExpandoObject using reflection to iterate over the properties in the specified class of type <see cref="T"/>.
+        /// A 
```
this needs cleaned up I guess

BrownMark commented on AdminCore.DataETL/TypeConverters/ListToStringConverter.cs:
```diff
@@ -0,0 +1,29 @@
+
+        public object ConvertTo(object value)
+        {
+            if (value == null)
```
if (value == null || !value.Count.Any()) {
return _defaultIfEmpty;
}

rgillen-uno replied:
I haven't combined theses two statements into the one conditional because I think I need to retain the usage of the object type when ConvertTo is called in CsvChoEtlAdapter. 
So I can't reference Count until I've checked for null then cast to List.

What do you think?

I did make a change further down to not use Count == 0.
```
if (!list.Any())
            {
                return _defaultIfNull;
            }
```

rgillen-uno replied:
I pushed up another commit for this. I realised that I was already doing type conversion in the automapper mapping, so i did it all in there and removed the TypeConverter attributes. I think it's maybe better to offload it onto an AutoMapper profile where you can be specific about casts rather than the generic object casting.

kkuczkowski-unosq commented on AdminCore.Services/EventService.cs:
```diff
@@ -77,16 +77,21 @@ public IList<EventDto> GetEventByStatus(EventStatuses eventStatus, EventTypes ev
       return _mapper.Map<IList<EventDto>>(events);
     }
 
+//    public string GetEventByStatusCsv(EventStatuses eventStatus, EventTypes eventType)
```
is this needed?

BrownMark commented on AdminCore.DataETL/TypeConverters/ListToStringConverter.cs:
```diff
@@ -0,0 +1,29 @@
+            {
+                return _defaultIfNull;
+            }
+            return string.Join($"{Environment.NewLine}", list);
```
Shouldn't need to interpolate the new line.

BrownMark commented on AdminCore.WebApp/Controllers/EventController.cs:
```diff
@@ -115,6 +119,16 @@ public IActionResult GetEventByStatusType(int eventStatusId, int eventTypeId)
+      var mappedEvents = _mapper.Map<IList<EventDataTransformModel>>(events);
+
+      var stream = new MemoryStream(_csvService.Generate(mappedEvents));
+      return File(stream, "application/octet-stream", "Report.csv");
```
I'd include the date in the name.

kkuczkowski-unosq replied:
can we figure out a way to make report names meaningful?

rgillen-uno replied:
I've pushed a commit that now gives them names such as "AwaitingApproval_AnnualLeave_Report13-05-2019.csv"

kkuczkowski-unosq replied:
that's perfect, thanks :+1:

BrownMark commented on AdminCore.Services/EventService.cs:
```diff
@@ -77,16 +77,21 @@ public IList<EventDto> GetEventByStatus(EventStatuses eventStatus, EventTypes ev
     }
 
+//    public string GetEventByStatusCsv(EventStatuses eventStatus, EventTypes eventType)
+//    {
```
Just need to clean this up.

BrownMark commented on AdminCore.DataETL/TypeConverters/TypeToStringConverter.cs:
```diff
@@ -0,0 +1,14 @@
+    {
+        public object ConvertTo(object value)
+        {
+            string enumValue = ((T)value)?.ToString();
```
This doesn't seem coupled to an enum.

BrownMark commented on AdminCore.DataETL/CsvChoEtlAdapter.cs:
```diff
@@ -0,0 +1,117 @@
+        /// <returns></returns>
+        private ExpandoObject CreateExpandoObject<T>(T obj)
+        {
+            Type type = obj.GetType();
```
I would use the var keyword for consistency. Makes refactoring easier too.

kkuczkowski-unosq commented on AdminCore.DataETL/CsvChoEtlAdapter.cs:
```diff
@@ -0,0 +1,117 @@
+        private List<ExpandoObject> CreateExpandoObjectList<T>(IList<T> data)
+        {
+            var objectsToWrite = new List<ExpandoObject>();
+            foreach (var dataItem in data)
```
I think this could be a LINQ-expression instead:
```
objectsToWrite.AddRange(data.Select(CreateExpandoObject));
```

kkuczkowski-unosq commented on AdminCore.DataETL/CsvChoEtlAdapter.cs:
```diff
@@ -0,0 +1,117 @@
+        /// <returns></returns>
+        private ChoCSVRecordConfiguration CreateConfig<T>()
+        {
+            ChoCSVRecordConfiguration config = new ChoCSVRecordConfiguration();
```
store as `var`?

BrownMark reviewed (changes requested):
Some initial comments.

rgillen-uno committed:
Remove commented code

kkuczkowski-unosq commented on AdminCore.DataETL/CsvChoEtlAdapter.cs:
```diff
@@ -0,0 +1,117 @@
+
+namespace AdminCore.DataETL
+{
+    public class CsvChoEtlAdapter : IDataEtlAdapter
```
what's `cho`? could we come up with a more meaningful name for this class?

rgillen-uno replied:
ChoEtl is the name of the ETL library https://github.com/Cinchoo/ChoETL
I think the name refers to it well enough unless you think differently?

kkuczkowski-unosq replied:
I don't have a better idea

rgillen-uno replied:
I've added an xml documentation comment to the class name describing purpose/implementation.
```
    /// <summary>
    /// Implements the ETL Library ChoEtl for conversion of data to csv.
    /// https://github.com/Cinchoo/ChoETL
    /// </summary>
```

rgillen-uno committed:
Use LINQ statement instead of foreach and use var n several places.

rgillen-uno committed:
Add DateService to EventController to allow adding correct date to ammended csv filename

rgillen-uno committed:
Remove string interpolation for newline environment variable.
Use linq .Any() method instead of .Count == 0.
Add xml doc comment to CsvChoEtlAdapter class name giving background on class purpose/implementation.
Rename of string 'enumValue' to mor meaningful name 'stringValue' as enums are now unrelated.

kkuczkowski-unosq reviewed (approved):
Given appropriate test coverage is added.

rgillen-uno committed:
Add null coalescing to return statement to return an empty string if null.

rgillen-uno committed:
LM-3: Add new tests for DataETL project.
Correct namespaces for common DataETL attributes.

rgillen-uno committed:
LM-3: Add some extra tests for CsvChoEtlAdapter

rgillen-uno committed:
Merge remote-tracking branch 'origin' into LM-3-file-export-leave-list

kkuczkowski-unosq commented on AdminCore.WebApp/Models/DataTransform/EventDataTransformModel.cs:
```diff
@@ -0,0 +1,34 @@
+{
+    public class EventDataTransformModel
+    {
+        [ExportableRecordField(Name = "Event Id", ColumnPosition = 1)]
```
should these be constants? maybe shared constants? thinking out loud.

rgillen-uno replied:
Which part specifically? The 'Name' parameter?

kkuczkowski-unosq replied:
yep, as discussed offline

rgillen-uno replied:
Decided to keep these attributes the way they are and not store 'Name' in as constants.

kkuczkowski-unosq commented on AdminCore.Services/Configuration/DependencyInjection.cs:
```diff
@@ -26,7 +27,7 @@ public static class DependencyInjection
     {
       if (!_registered)
-      {
+      {        
```
whitespace, there are some other instances but I won't comment on each

kkuczkowski-unosq commented on AdminCore.DataETL/TypeConverters/TypeToStringConverter.cs:
```diff
@@ -0,0 +1,16 @@
+        public object ConvertTo(object value)
+        {
+            string stringValue = ((T) value)?.ToString();
+            return stringValue ?? "";
```
I'm missing a lot of context but why return an empty string? This implies the value returned was available which is not true

rgillen-uno replied:
My reasoning is that if a null value is converted that should be represented in the CSV/exported file as an empty string rather than "null".

kkuczkowski-unosq replied:
OK 🤔

kkuczkowski-unosq commented on AdminCore.DataETL.Tests/Models/DataETLTestModelSimpleDuplicateRecordField.cs:
```diff
@@ -0,0 +1,14 @@
+{
+    public class DataETLTestModelSimpleDuplicateRecordField
+    {
+        [ExportableRecordField(Name = "Field1", ColumnPosition = 1)]
```
store this as constant and reuse?

rgillen-uno replied:
As discussed, no change

kkuczkowski-unosq commented on AdminCore.DataETL.Tests/ClassData/CsvChoEtlAdapterClassData.cs:
```diff
@@ -0,0 +1,91 @@
+        {
+            public IEnumerator<object[]> GetEnumerator()
+            {
+                // List<DataETLTestModelListToStringConverter>, expectedCsv
```
this needs cleaned up?

rgillen-uno committed:
LM-3: Remove whitespace

kkuczkowski-unosq commented on AdminCore.DataETL.Tests/ClassData/CsvChoEtlAdapterClassData.cs:
```diff
@@ -0,0 +1,91 @@
+                            StringCollection = new List<string> { "A", "B", "C" }
+                        }   
+                    },
+                    "Field1\r\n\"A\r\nB\r\nC\""
```
I would use a constant in this instance, give this string a meaningful name. This would aid readability IMO

rgillen-uno replied:
That's a good idea, I'll implement this now

kkuczkowski-unosq commented on AdminCore.DataETL.Tests/ClassData/CsvChoEtlAdapterClassData.cs:
```diff
@@ -0,0 +1,91 @@
+                            DataETLTestEnum1 = 2
+                        }   
+                    },
+                    "Field1\r\ncEnum"
```
same here with the constant

kkuczkowski-unosq commented on AdminCore.DataETL.Tests/ClassData/CsvChoEtlAdapterClassData.cs:
```diff
@@ -0,0 +1,91 @@
+        {
+            public IEnumerator<object[]> GetEnumerator()
+            {
+                // List<DataETLTestModelTypeToStringConverter>, expectedCsv
```
clean up?

rgillen-uno replied:
Those are the arguments that need passed into the test method but I'll make that clearer

kkuczkowski-unosq commented on AdminCore.DataETL.Tests/ClassData/CsvChoEtlAdapterClassData.cs:
```diff
@@ -0,0 +1,91 @@
+                            Field_DateTime = new DateTime(1970, 1, 1)
+                        }   
+                    },
+                    "Field1,Field2,Field3,Field4,Field5\r\n" +
```
consider using a constant

rgillen-uno committed:
LM-3: Add constants in place of strings

rgillen-uno committed:
LM-3: Removed TypeConverter attributes and do all conversion in AutoMapper Profile.
Renamed DataETL classes to something more meaningful.
Removed redundant tests.
Added new tests.

rgillen-uno committed:
LM-3: Change octet magic string to enum.

rgillen-uno committed:
LM-3: Add new Ignore field attribute and appropriate tests

rgillen-uno committed:
Merge branch 'develop' into LM-3-file-export-leave-list
